File: java/05-reactor/src/test/java/com/fResult/reactor/ch5_10/ControlFlowTimeoutTest.java
package com.fResult.reactor.ch5_10;

import static org.junit.jupiter.api.Assertions.assertInstanceOf;

import java.time.Duration;
import java.util.concurrent.TimeoutException;
import lombok.extern.log4j.Log4j2;
import lombok.val;
import org.junit.jupiter.api.Test;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;
import reactor.test.StepVerifier;

@Log4j2
class ControlFlowTimeoutTest {
  @Test
  void timeout() {
    val ids =
        Flux.just(1, 2, 3)
            .delayElements(Duration.ofSeconds(1))
            .timeout(Duration.ofMillis(500))
            .onErrorResume(this::fallbackOnTimeout);

    StepVerifier.create(ids).expectNext(0).verifyComplete();
  }

  private Publisher<? extends Integer> fallbackOnTimeout(Throwable ex) {
    log.warn("Timeout occurred: {}", ex.getMessage());
    assertInstanceOf(
        TimeoutException.class,
        ex,
        "Expected TimeoutException, but got: " + ex.getClass().getSimpleName());
    return Flux.just(0);
  }
}

File: kotlin/05-reactor/src/test/kotlin/com/fResult/reactor/ch5_10/ControlFlowTimeoutTest.kt
package com.fResult.reactor.ch5_10

import org.apache.logging.log4j.LogManager
import reactor.core.publisher.Flux
import reactor.test.StepVerifier
import java.util.concurrent.TimeoutException
import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

class ControlFlowTimeoutTest {
  private val log = LogManager.getLogger(ControlFlowTimeoutTest::class.java)

  @Test
  fun timeout() {
    val ids =
      Flux.just(1, 2, 3)
        .delayElements(1.seconds.toJavaDuration())
        .timeout(500.milliseconds.toJavaDuration())
        .onErrorResume(::fallbackOnTimeout)

    StepVerifier.create(ids).expectNext(0).verifyComplete()
  }

  private fun fallbackOnTimeout(ex: Throwable): Flux<Int> {
    log.warn("Timeout occurred: {}", ex.message)
    assertTrue(ex is TimeoutException, "Expected TimeoutException, but got ${ex::class.simpleName}")
    return Flux.just(0)
  }
}
